common/net/credentials/alts/handshaker/handshaker.go

```diff
@@ -0,0 +1,373 @@
+	return &altsHandshaker{
+		protocol:   recordxChaChaPolyProtocol,
+		conn:       conn,
+		serverOpts: opts,
+		side:       alts.ServerSide,
+	}, nil
+}
+
+// read and decode the handshake record
+func (s *altsHandshaker) readHandshake(value interface{}) error {
+	hdr := make([]byte, handshakeHeader)
+	// read handshake header from connection
+	if _, err := s.conn.Read(hdr); err != nil {
+		return fmt.Errorf("read handshake header: %w", err)
+	}
+
+	// the handshake body length
+	n := int(hdr[0])<<8 | int(hdr[1])
+
+	body := make([]byte, n)
+	// read handshake body from connection
+	if _, err := s.conn.Read(body); err != nil {
+		return fmt.Errorf("read handshake body: %w", err)
+	}
+
+	// new a gob decoder for the handshake record
+	decoder := gob.NewDecoder(bytes.NewBuffer(body))
+	// decode the handshake record
+	if err := decoder.Decode(value); err != nil {
+		return fmt.Errorf("decode handshake record: %w", err)
+	}
+	return nil
+}
+
+// encode and write the handshake record
+func (s *altsHandshaker) writeHandshake(value interface{}) error {
+	var buf bytes.Buffer
+	// new a gob encoder for the handshake record
+	encoder := gob.NewEncoder(&buf)
+
+	// encode the handshake record
+	if err := encoder.Encode(value); err != nil {
+		return fmt.Errorf("encode handshake record: %w", err)
+	}
+
+	m := buf.Len()
+	d := make([]byte, handshakeHeader+m)
+	d[0] = byte(m >> 8)
+	d[1] = byte(m)
+	copy(d[handshakeHeader:], buf.Bytes())
+
+	// write the handshake record
+	if _, err := s.conn.Write(d); err != nil {
+		return fmt.Errorf("write handshake record: %w", err)
+	}
+	return nil
+}
+
+func (s *altsHandshaker) doClientHandshake(ctx context.Context, auth alts.Authentication, signInfo *alts.SignInfo) error {
+	curve := ed448.NewCurve()
+	// generate the private and public key for client
+	priv, pub, ok := curve.GenerateKeys()
+	if !ok {
+		return errors.New("failed to generate keys")
+	}
+
+	signature, err := auth.Sign(ctx, pub[:], signInfo.PastelID, signInfo.PassPhrase)
+	if err != nil {
+		return fmt.Errorf("failed to generate signature: %w", err)
+	}
+
+	request := kexExchangeRequest{
+		PubKey:    pub[:],
+		Signature: signature,
+		PastelID:  signInfo.PastelID,
+	}
+	// encode and write the client's handshake record
+	if err := s.writeHandshake(&request); err != nil {
+		return fmt.Errorf("write handshake: %w", err)
+	}
+
+	// read and decode the client's handshake record
+	var response kexExchangeResponse
+	if err := s.readHandshake(&response); err != nil {
+		return fmt.Errorf("read handshake: %w", err)
+	}
+
+	if ok, err := auth.Verify(ctx, response.PubKey, string(response.Signature), response.PastelID); err != nil || !ok {
+		return fmt.Errorf("failed to verify server public key: %w", err)
+	}
+
+	var serverPubKey [ED448PubKeySize]byte
+	copy(serverPubKey[:], response.PubKey)
+	// compute the secret key for connection
+	secret := curve.ComputeSecret(priv, serverPubKey)
+
+	subkeyLen, ok := recordKeyLen[s.protocol]
+	if !ok {
```

Comment:
please ensure len(secret) > cryptokdf.CryptoKdfKeybytes(). should validate it here.